src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs

```diff
@@ -277,25 +270,26 @@ async Task ProcessFileAndComplete(ILearningTransportTransaction transaction, str
 
                 try
                 {
-                    await onCompleted(new CompleteContext(messageId, wasAcknowledged, result.Headers, startedAt, DateTimeOffset.UtcNow, result.ProcessingFailed, processingContext), messageProcessingCancellationToken).ConfigureAwait(false);
+                    await onCompleted(new CompleteContext(messageId, wasAcknowledged, headers ?? new Dictionary<string, string>(), startedAt, DateTimeOffset.UtcNow, onMessageFailed, processingContext), messageProcessingCancellationToken).ConfigureAwait(false);
                 }
                 catch (Exception ex)
                 {
-                    log.Warn($"Failure when invoking OnComplete for {filePath}({transaction.FileToProcess})" + filePath, ex);
+                    log.Warn($"Failure when invoking {nameof(OnCompleted)} for {filePath}({transaction.FileToProcess})" + filePath, ex);
                 }
 
                 concurrencyLimiter.Release();
             }
         }
 
-        async Task<ProcessFileResult> ProcessFile(ILearningTransportTransaction transaction, string messageId, ContextBag processingContext, CancellationToken messageProcessingCancellationToken)
+        /// <returns>A <see cref="Dictionary{TKey, TValue}"/> of headers, and a <see cref="bool"/> indicating whether onMessage failed.</returns>
+        async Task<(Dictionary<string, string>, bool)> ProcessFile(ILearningTransportTransaction transaction, string messageId, ContextBag processingContext, CancellationToken messageProcessingCancellationToken)
```

Comment:
If the tuple is not to taste, we can re-introduce a type, but I'm not sure it should be a `class`, since that introduces an allocation per message. Then again, this is only the learning transport.